packages/async-flow/src/replay-membrane.js

```diff
@@ -0,0 +1,389 @@
+   * @param {number} callIndex
+   * @param {Host} hostProblem
+   * @returns {Outcome}
+   */
+  const doThrow = (callIndex, hostProblem) => {
+    unnestInterpreter(callIndex);
+    const guestProblem = hostToGuest(hostProblem);
+    return harden({
+      kind: 'throw',
+      problem: guestProblem,
+    });
+  };
+
+  // ///////////// Guest to Host or consume log ////////////////////////////////
+
+  const performCall = (hostTarget, optVerb, hostArgs, callIndex) => {
+    let hostResult;
+    try {
+      hostResult = optVerb
+        ? hostTarget[optVerb](...hostArgs)
+        : hostTarget(...hostArgs);
+    } catch (hostProblem) {
+      return logDo(nestDispatch, harden(['doThrow', callIndex, hostProblem]));
+    }
+    return logDo(nestDispatch, harden(['doReturn', callIndex, hostResult]));
+  };
+
+  const guestCallsHost = (guestTarget, optVerb, guestArgs, callIndex) => {
+    if (!bijection.hasGuest(guestTarget)) {
+      // This happens in a delayed guest-to-host call from a previous run.
+      // In that case, the bijection was reset and all guest caps
+      // created in the previous run were unregistered,
+      // including guestTarget.
+      // Throwing an error back to the old guest caller may cause
+      // it to proceed in all sorts of crazy ways. But that old run
+      // should now be isolated and unable to cause any observable effects.
+      // Well, except for resource exhaustion including infinite loops,
+      // which would be a genuine problem.
+      //
+      // Console logging of unhandled rejections, errors thrown to the top
+      // of the event loop, or anything else are not problematic effects.
+      // At this level of abstraction, we don't consider console logging
+      // activity to be observable. Thus, it is also ok for the guest
+      // function, which should otherwise be closed, to
+      // capture (lexically "close over") the `console`.
+      const extraDiagnostic =
+        callStack.length === 0
+          ? ''
+          : // This case should only happen when the callStack is empty
+            ` with non-empty callstack ${q(callStack)};`;
+      Fail`Called from a previous run: ${guestTarget}${b(extraDiagnostic)}`;
+    }
+    /** @type {Outcome} */
+    let outcome;
+    try {
+      const guestEntry = harden([
+        'checkCall',
+        guestTarget,
+        optVerb,
+        guestArgs,
+        callIndex,
+      ]);
+      if (log.isReplaying()) {
+        const entry = log.nextEntry();
+        equate(
+          guestEntry,
+          entry,
+          `replay ${callIndex}:
+     ${q(guestEntry)}
+  vs ${q(entry)}
+    `,
+        );
+        outcome = /** @type {Outcome} */ (nestInterpreter(callIndex));
+      } else {
+        const entry = guestToHost(guestEntry);
+        log.pushEntry(entry);
+        const [_, ...args] = entry;
+        nestInterpreter(callIndex);
+        outcome = performCall(...args);
+      }
+    } catch (fatalError) {
+      return panic(fatalError);
```

Comment:
Why not throw here?
```suggestion
      throw panic(fatalError);
```

I'm actually unsure either is right. Really what we want is to return something that will prevent the guest from making any further forward progress if possible.

Reply:
> I'm actually unsure either is right. Really what we want is to return something that will prevent the guest from making any further forward progress if possible.

IIUC, there is nothing we could return or throw that would prevent the guest from making further progress. If I'm missing something, please reopen.

I used to have `panic` return `undefined` but now I do have it always throw. The `return` above was vestigial. Fixed.

The best we can do with post-revoke behavior of the guest code is to prevent it from doing any damage, under the assumption that it was not written to do any damage. AFAICT, the best we can do is to address the following issues, which are good enough when the guest is assumed not to be attacking the host
- Even a benign guest should be assumed to be written naively to execution after revocation. Thus, once revoked, a naive guest may "go crazy", acting in ways outside its author's expectations. However, JS code is generally written assuming any call might throw rather than return. Now that a revoked membrane always throws, this should restore the assumptions made by the author of even a naive guest function.
- If the author has not thought about the throw case, then they probably do not catch the throw but let it propagate, thus effectively terminating the guest execution immediately. If they do catch it with a `try/catch`, then they probably have put some thought into being throw-tolerant.
- The membrane prevents any integrity effects from guest to host, effectively isolating the guest re integrity. For these purposes, there is no confidentiality concern. The remaining "go crazy" issue would be unreasonable resource usage, such as an infinite loop.

If I've missed any remaining dangers for post-revocation guest functions, please reopen and explain. Attn @raphdev @LuqiPan . Thanks. Resolving